x-pack/plugins/graph/public/helpers/use_workspace_loader.ts

```diff
@@ -71,24 +84,43 @@ export const useWorkspaceLoader = ({
         .then((response) => response.savedObjects);
     }
 
-    async function fetchSavedWorkspace() {
-      return (id
+    async function fetchSavedWorkspace(): Promise<{
+      savedObject: GraphWorkspaceSavedObject;
+      sharingSavedObjectProps?: SharingSavedObjectProps;
+    }> {
+      return id
         ? await getSavedWorkspace(savedObjectsClient, id).catch(function (e) {
-            toastNotifications.addError(e, {
+            coreStart.notifications.toasts.addError(e, {
               title: i18n.translate('xpack.graph.missingWorkspaceErrorMessage', {
                 defaultMessage: "Couldn't load graph with ID",
               }),
             });
-            history.replace('/home');
+            historyReplace('/home');
             // return promise that never returns to prevent the controller from loading
             return new Promise(() => {});
           })
-        : await getSavedWorkspace(savedObjectsClient)) as GraphWorkspaceSavedObject;
+        : getEmptyWorkspace();
     }
 
     async function initializeWorkspace() {
       const fetchedIndexPatterns = await fetchIndexPatterns();
-      const fetchedSavedWorkspace = await fetchSavedWorkspace();
+      const {
+        savedObject: fetchedSavedWorkspace,
+        sharingSavedObjectProps: fetchedSharingSavedObjectProps,
+      } = await fetchSavedWorkspace();
+
+      if (spaces && fetchedSharingSavedObjectProps?.outcome === 'aliasMatch') {
+        // We found this object by a legacy URL alias from its old ID; redirect the user to the page with its new ID, preserving any URL hash
+        const newObjectId = fetchedSharingSavedObjectProps?.aliasTargetId!; // This is always defined if outcome === 'aliasMatch'
+        const newPath = getEditUrl(coreStart.http.basePath.prepend, { id: newObjectId });
+        spaces.ui.redirectLegacyUrl(
+          newPath,
```

Comment:
Here we need to provide the search query if have one.

```suggestion
        const newURLQuery = newPath + search;
        spaces.ui.redirectLegacyUrl(
          newURLQuery,
```